File: app/src/main/java/com/example/test/MainConverter.java
package com.example.test;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class MainConverter {

    @Inject
    public MainConverter() {
    }

    public List<CardInfoItemModel> convert(Card card) {
        List<CardInfoItemModel> listCardInfo = new ArrayList<>();
        if (infoRealValue(card.getScheme())) {
            MainCardInfo mainCardInfoScheme = new MainCardInfo("Scheme:", card.getScheme());
            listCardInfo.add(mainCardInfoScheme);
        }
        if (infoRealValue(card.getType())) {
            MainCardInfo mainCardInfoType = new MainCardInfo("Type:", card.getType());
            listCardInfo.add(mainCardInfoType);
        }
        if (infoRealValue(card.getBrand())) {
            MainCardInfo mainCardInfoBrand = new MainCardInfo("Brand:", card.getBrand());
            listCardInfo.add(mainCardInfoBrand);
        }
        MainCardInfo mainCardInfoPrepaid = new MainCardInfo("Prepaid:", booleanToString(card.isPrepaid(), "Yes", "No"));
        listCardInfo.add(mainCardInfoPrepaid);
        if (infoRealValue(card.getCountry().getCurrency())) {
            MainCardInfo mainCardInfoBrand = new MainCardInfo("Currency:", card.getCountry().getCurrency());
            listCardInfo.add(mainCardInfoBrand);
        }
        if (numberInfo(card.getNumber()) != null) {
            MainCardInfo mainCardInfoNumber = new MainCardInfo("Number:", numberInfo(card.getNumber()));
            listCardInfo.add(mainCardInfoNumber);
        }
        CountryInfoItemModel countryInfoItemModel = new CountryInfoItemModel(
                "Country name:",
                "Coordinate:",
                card.getCountry().getAlpha2() + " " + card.getCountry().getName(),
                card.getCountry().getLatitude() + "/" + card.getCountry().getLongitude());
        listCardInfo.add(countryInfoItemModel);
        if (card.getBank().getName() != null || card.getBank().getUrl() != null || card.getBank().getPhone() != null) {
            BankInfoItemModel bankInfoItemModel = new BankInfoItemModel(
                    "Bank info:",
                    card.getBank().getName(),
                    card.getBank().getUrl(),
                    card.getBank().getPhone()
            );
            listCardInfo.add(bankInfoItemModel);
        }
        return listCardInfo;

    }

    private String numberInfo(Number number) {
        String numberCardInfo = null;
        if (number.getLength() != null && number.isLuhn() != null) {
            numberCardInfo = "Length: " + number.getLength() + "\n" + "Luhn: " + booleanToString(number.isLuhn(), "Yes", "No");
        } else if (number.getLength() != null) {
            numberCardInfo = "Length: " + number.getLength();
        } else if (number.isLuhn() != null) {
            numberCardInfo = "Luhn: " + booleanToString(number.isLuhn(), "Yes", "No");
        }
        return numberCardInfo;
    }

    private String booleanToString(boolean predicate, String trueString, String falseString) {
        String result;
        if (predicate) {
            result = trueString;
        } else {
            result = falseString;
        }
        return result;
    }

    private boolean infoRealValue(String info) {
        boolean infoIsNotEmpty = false;
        if (info != null && !info.isEmpty()) {
            infoIsNotEmpty = true;
        }
        return infoIsNotEmpty;
    }
}

File: app/src/main/java/com/example/test/data/api/MainConverter.kt
package com.example.test.data.api

import com.example.test.data.model.CardDto
import com.example.test.data.model.NumberDto
import com.example.test.domain.model.BankInfoItemModel
import com.example.test.domain.model.CardInfoItemModel
import com.example.test.domain.model.CountryInfoItemModel
import com.example.test.domain.model.MainCardInfo
import javax.inject.Inject

class MainConverter @Inject constructor() {

    fun convert(card: CardDto): List<CardInfoItemModel> {
        val listCardInfo: MutableList<CardInfoItemModel> = mutableListOf()
        if (infoRealValue(card.scheme)) {
            val mainCardInfoScheme: MainCardInfo = MainCardInfo("Scheme:", card.scheme)
            listCardInfo.add(mainCardInfoScheme)
        }
        if (infoRealValue(card.type)) {
            val mainCardInfoType: MainCardInfo = MainCardInfo("Type:", card.type)
            listCardInfo.add(mainCardInfoType)
        }
        if (infoRealValue(card.brand)) {
            val mainCardInfoBrand: MainCardInfo = MainCardInfo("Brand:", card.brand)
            listCardInfo.add(mainCardInfoBrand)
        }
        val mainCardInfoPrepaid: MainCardInfo = MainCardInfo("Prepaid:", booleanToString(card.isPrepaid, "Yes", "No"))
        listCardInfo.add(mainCardInfoPrepaid)
        if (infoRealValue(card.country.currency)) {
            val mainCardInfoBrand: MainCardInfo = MainCardInfo("Currency:", card.country.currency)
            listCardInfo.add(mainCardInfoBrand)
        }
        if (numberInfo(card.number) != null) {
            val mainCardInfoNumber: MainCardInfo = MainCardInfo("Number:", numberInfo(card.number))
            listCardInfo.add(mainCardInfoNumber)
        }
        val countryInfoItemModel: CountryInfoItemModel = CountryInfoItemModel(
            "Country name:",
            "Coordinate:",
            "${card.country.alpha2} ${card.country.name}",
            "${card.country.latitude}/${card.country.longitude}")
        listCardInfo.add(countryInfoItemModel)
        if (card.bank.name != null || card.bank.url != null || card.bank.phone != null) {
            val bankInfoItemModel: BankInfoItemModel = BankInfoItemModel(
                "Bank info:",
                card.bank.name,
                card.bank.url,
                card.bank.phone
            )
            listCardInfo.add(bankInfoItemModel)
        }
        return listCardInfo
    }

    private fun numberInfo(number: NumberDto): String? {
        var numberCardInfo: String? = null
        if (number.length != null && number.isLuhn != null) {
            numberCardInfo = "Length: " + number.length + "\n" + "Luhn: " + booleanToString(number.isLuhn, "Yes", "No")
        } else if (number.length != null) {
            numberCardInfo = "Length: " + number.length;
        } else if (number.isLuhn != null) {
            numberCardInfo = "Luhn: " + booleanToString(number.isLuhn, "Yes", "No")
        }
        return numberCardInfo;
    }

    private fun booleanToString(predicate: Boolean, trueString: String, falseString: String): String {
        val result: String = if (predicate) {
            trueString
        } else {
            falseString
        }
        return result
    }

    private fun infoRealValue(info: String): Boolean {
        var infoIsNotEmpty: Boolean = false
        if (info.isNotEmpty()) {
            infoIsNotEmpty = true
        }
        return infoIsNotEmpty
    }
}